### DISCUSSION TITLE
What to do about Meson?

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7546] [Development]
I added Meson with the idea that we could gradually move to it from CMake, but I'm now having second thoughts. 

My experience of Meson so far is that it's a strict and opinionated system (which is not necessarily bad, but makes it time consuming for us to transition). I am considering Meson--in its current state--best for new projects, and maybe transitioning from smaller CMake projects; perhaps it doesn't make sense to transition a large CMake project to Meson.

Being able to do this kind of thing is nice:
```
system_cli11 = get_option('system-cli11')
if system_cli11
  dependency('cli11', required: false)
else
  subproject('cli11')
endif
```

But perhaps CMake [`FetchContent`](https://cmake.org/cmake/help/latest/module/FetchContent.html) is better for us right now. Our [Meson config](https://github.com/deskflow/deskflow/blob/4256ea2fbe561d74ae3470c04b66f45c118227ba/meson.build) is currently small and only satisfies dependencies (with `wrap-file` and `wrap-git`), so it's probably the best opportunity to back out.

**Comment by sithlord48** [comment]
+1  Cmake

**Reply by nbolton** [reply]
Alright, let's start removing Meson. Unless anyone objects?

**Reply by enzious** [reply]
I am not an experienced C++ developer, so, what were your motivations for trying to use meson.

**Reply by nbolton** [reply]
Yeah, I didn't really mention that, did I? 

Ok, so I was trying to pull in unreleased changes from [`libportal`](https://github.com/flatpak/libportal) as they were taking a little while to release 0.8 and @whot had made some pretty brilliant changes to that lib which we wanted to use. Initially, I attempted to use CMake `FetchContent` but found I had to also run extra meson commands to build `libportal`, which meant making sure meson was installed, etc... then I thought, "Hey, why not just use meson to pull in and build the meson project as a subproject... I can also do the same for `libei`"... and at that point I got carried away and started adding all sorts. Now I'm starting to feel some regret and want to back out the meson stuff, then try again with CMake.

**Reply by Conan-Kudo** [reply]
I personally prefer CMake over Meson, as I find Meson's lack of flexibility and extensibility to be pretty painful for non-trivial projects. And the lack of QML support sure doesn't help.

**Comment by whot** [comment]
regarding the snippet, meson is actually smarter here and you can do:
```
dependency('cli11', required: get_option('system-cli11'))
```
and if that's a feature it'll handle auto/enabled/disabled correctly. works for boolean too.

There's also `subdir_done()` which you can use to keep higher-level meson files clean - you enter the subdir unconditionally and then exit based on some condition (e.g. windows vs linux check).

and for subprojects you can use those as `fallback: ...` in the `dependency()` function - to be used where the system one isn't available.

**Reply by nbolton** [reply]
Good tip. That looks cleaner and simpler. I hadn't come across `subdir_done()`, I'll try to use that in future.

**Comment by whot** [comment]
> my experience of Meson so far is that it's a strict and opinionated system

meson by and large works on the principle of "you're not that special" and tries to solve things in a way that applies to other projects that have the same requirements. This makes some things a bit awkward but in most cases you just need to find the way it deals with things and then it becomes quite trivial to maintain.

That's not a "you must switch to meson" but so far every project I have switched has been a long-term benefit in simplified maintenance. Never did anything using macos or windows though and I never switched a cmake projects, so take the above with that caveat.

**Reply by nbolton** [reply]
I agree. It's about adapting to the Meson-way, not about trying to get Meson to do it your way.

> That's not a "you must switch to meson" but so far every project I have switched has been a long-term benefit in simplified maintenance

What I might do is learn Meson on a smaller project first, then, if we do decide to switch (I'm open-minded about it), we can make sure we do it properly.

So far I am beginning to see that transitioning piecemeal from CMake to Meson is perhaps not the right approach since you have to implement everything twice (once in Meson and once in CMake). The switch seems to be all or nothing.

**Reply by whot** [reply]
> meson is perhaps not the right approach since you have to implement everything twice (once in Meson and once in CMake). The switch seems to be all or nothing.

mixing build systems almost never works well - it's possible to support both (for a while anyway) but mixing them is not for the faint-hearted. None of the build systems are really designed for that.

**Reply by nbolton** [reply]
Learned that the hard way. I'll try to revert to CMake alone for 1.18.

**Comment by nbolton** [comment]
Raised: #7679
